pkg/kf/internal/routeutil/routeutil.go

```diff
@@ -0,0 +1,40 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package routeutil
+
+import (
+	"encoding/base64"
+	"strings"
+)
+
+// EncodeRouteName is used to create a valid DNS name that can be used as a
+// name for something like a VirtualService.
+func EncodeRouteName(hostname, domain, urlPath string) string {
+	return strings.ToLower(base64.RawURLEncoding.EncodeToString([]byte(hostname + domain + urlPath)))
```

Comment:
Lowering this could lead to collisions in the namespace.

It might make sense to somehow take the hash of the three then convert the 64 bit hash to base64 to be something like `$hostname-base64($HASH)`, Knative does that with their resources. That ensures they're always a valid length and are mostly resistant to collisions.

Reply:
I updated it, PTAL.